### PR TITLE
Robertpangborn/io tversions

### PR DESCRIPTION
**Pull request by robertpangborn**

### CODE REVIEW

File: container/dockershell.sh [hunk 1]
```diff
@@ -145,6 +150,11 @@ if [ -z ${DEPLOYMENT_TYPE} ]; then
     DEPLOYMENT_TYPE=$defaultDeployType
 fi
 
+if [ -z ${ES_VERSION} ]; then
+    printf "Event Store version (--es-version) not specified, please provide"
```
Comment (cgarciaarellano) [inline container/dockershell.sh]: I would indicate that we will use the default version (even mentioning that version explicitly), rather than saying "please provide"

File: container/Dockerfile [hunk 2]
```diff
@@ -37,19 +36,11 @@ RUN echo -e 'startup_message off \nhardstatus on \nhardstatus alwayslastline \nv
 # Setup vim 
 RUN echo -e 'syntax enable \nset tabstop=4 \nset number \nset hlsearch \nset incsearch' > ${HOME}/.vimrc
 
-# Setup spark/scala/python/rest
+# Setup rest
```
Comment (hroumani) [inline container/Dockerfile]: Can you move this comment to above line 42, it's more appropriate there.

File: container/dockershell.sh [hunk 3]
```diff
@@ -145,6 +150,11 @@ if [ -z ${DEPLOYMENT_TYPE} ]; then
...
+    DEPLOYMENT_TYPE=$defaultDeployType
```
Comment (cgarciaarellano) [inline container/dockershell.sh]: hmm is this line necessary here ? how is this related to the ES_VERSION ?

File: container/dockershell.sh [hunk 4]
```diff
@@ -57,6 +57,7 @@ OPTIONS:
                   e.g. "db2eventstore-1578174815082"
 --user         User name of the Event Store server
 --password     Password of the Event Store server
+--es-version   The Event Store version you want to install
```
Comment (hroumani) [inline container/dockershell.sh]: I think this is a little confusing as it's not the eventstore version you want to install but rather:
`The event store version of the target cluster`

File: container/dockershell.sh (hunk 1 above)
Comment (hroumani) [inline container/dockershell.sh]: No we want this to default to latest though ... otherwise they will always need to supply a version here.

This should be seamless for most users who do not care as they're using the latest release, so we should default to that.

File: container/dockershell.sh (hunk 3 above)
Comment (hroumani) [reply container/dockershell.sh]: +1 I think this was a typo / copy and paste error.

File: container/dockershell.sh [hunk 5]
```diff
@@ -171,17 +181,23 @@ fi
 mkdir -p ${USER_VOLUME}
 # start container in interactive mode
 
-entryPoint="env && ${SETUP_PATH}/setup-ssl.sh && ${SETUP_PATH}/entrypoint_msg.sh && bash --login"
+echo "setting the entrypoint variable"
+echo $SETUP_PATH
```
Comment (hroumani) [inline container/dockershell.sh]: What's the point of printing this, I would just delete it.

File: container/dockershell.sh [hunk 6]
```diff
@@ -171,17 +181,23 @@ fi
...
+entryPoint="env && ${SETUP_PATH}/setup-ssl.sh && ${SETUP_PATH}/setup-container.sh ${ES_VERSION} && ${SETUP_PATH}/entrypoint_msg.sh && bash --login"
```
Comment (cgarciaarellano) [inline container/dockershell.sh]: Hmm, I see why you were concerned about the time it took to download the libraries, I had not understood that we would set this up at runtime as the container was started up each time. @hroumani could we discuss this ?

File: container/es-releases.json [hunk 7]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (hroumani) [inline container/es-releases.json:1]: This looks good, nice.

**Review by hroumani — CHANGES_REQUESTED** [review]
So a few changes requested.

In addition, I'd like to do a few tests on my own.  We decided to move this within the container each time it's run to allow us to dynamically change versions (e.g. scenario being user built image, and then using same image pointed to multiple eventstore clusters of differing versions).

But this was done **under the premise** that the setup was fairly quick (i.e. the steps in `setup-container.sh`, particularly with the removal of kafka).  But there were a few reports that this is in fact not the case.  I have a suggestion to handle this I'll make but we don't necessarily need to go that route if this is indeed fairly quick.

**Comment by hroumani** [comment]
FYI @robertpangborn 

Assuming `setup-container` now invoked in the entrypoint i.e. container startup is slow ~ 1min+ I have the following suggestion in order for us to still support avoiding an image rebuild each time a user wants to target a different version:

```
- Add a call to setup-container.sh in dockerfile i.e. the image (note you'll need to COPY the JSON file in the dockerfile, this will be in a different place from ~/user_volume i.e. just for the image build step)

- Now store the version number used (based on the input to image build) in a file in a well known location (e.g imageBuild.esversion).

- In addition setup-container.sh will be modified such that an additional step is added that checks for the existence of the version file mentioned above
- If said file is present, the md5sum will be calculated of the ~/user_volume (i.e. the version of the json file supplied each time the container is started) and compared to the file ... if they're the same, no setup is needed because the default setup done during image build is still fine ...
```

Please note ... this will not help a user who is often changing the version i.e. if the md5sum differs, it will always differ everytime the container starts, as of course everything we do in the entrypoint is ephemeral. 

In addition if we go this route we can add kafka back and keep it only in the dockerfile (i.e. image build step) => we get that as well, this step we know takes quite a while so we would not want to run it dynamically.

But before we go this route I'd like to see how long it takes on average, I'll start it a few times.


Finally we would change the image build step so that you can specify the version of image build (i.e. just like the same --es-version input to dockershell) => they can change this.

File: container/dockershell.sh (hunk 1 above)
Comment (robertpangborn) [reply container/dockershell.sh]: I am going to remove this as in the setup-container script i test if a version was specified and if not it uses 2.0.1.0. Which means this script will need to be updated when the json file is update with the next realese. 

Perhaps the docs should be explicit to which version is current and that is what the scripts will use?

**Review by hroumani — CHANGES_REQUESTED** [review]
Another point;

All the various apps do the following (as an example) to compile/run:
```
echo "Compile java code"
#1. download & install jdk 1.8<
#2. download & install scala 2.11.8<
#3. set SPARK_HOME = < spark 2.0.2 or spark 2.1 installation path>
ESLIB=${SPARK_HOME}/jars/ibm-db2-eventstore-client-spark-2.2.1-2.0.0.jar

echo "Compile the scala application"
rm -f myScalaExample.jar
scalac ExampleScalaApp.scala -d myScalaExample.jar \
>---classpath ${ESLIB}:${SPARK_HOME}/jars/*
```
You can see the JAR version is hard-coded, but in fact after this change, we will now use the JAR supplied by the client (based on es-version) => this can break if they specify a non-default version.

Simple fix though, just download it to a fixed name, and change all the scripts to use that instead :)  e.g. `ibm-db2-eventstore.jar`.

File: container/dockershell.sh (hunk 3 above)
Comment (robertpangborn) [reply container/dockershell.sh]: yes, that if test is proposed to be removed because version existance tested for in setup container

File: container/dockershell.sh (hunk 1 above)
Comment (hroumani) [reply container/dockershell.sh]: Sounds good, update the README and script instructions though to indicate what happens when it's not specified.

File: container/dockershell.sh (hunk 6 above)
Comment (hroumani) [reply container/dockershell.sh]: I touch on this in the comments below, we'll chat.

File: container/Dockerfile (hunk 2 above)
Comment (robertpangborn) [reply container/Dockerfile]: yes

File: container/dockershell.sh (hunk 1 above)
Comment (robertpangborn) [reply container/dockershell.sh]: resolved by setup-container supplying current release if no version supplied

File: container/dockershell.sh (hunk 4 above)
Comment (robertpangborn) [reply container/dockershell.sh]: I am not seeing the difference, but it might be not geting the signifigance of "target cluster" but can change it

File: container/dockershell.sh (hunk 5 above)
Comment (robertpangborn) [reply container/dockershell.sh]: those were added by me to help with debugging, have removed them

File: container/dockershell.sh (hunk 4 above)
Comment (hroumani) [reply container/dockershell.sh]: The point being `dockershell` and the `IOT application container` is not installing anything per se, rather it is setting up a client to point to an already installed eventstore cluster.

So the idea is to avoid any confusion that we are "installing eventstore" locally.

**Comment by hroumani** [comment]
We discussed this and decided will go this direction ... so we'll give up the flexibility of dynamically picking up new versions per docker run for a simpler approach at image build time only:

- `build.sh` will now also take an `--es-version` input parameter
- By default we will use this version for the `TAG` e.g. -> eventstore_demo:latest` or `eventstore_demo:2.0.0.5`
- We will now call `setup-container.sh` (i.e. pull/install the correct artifacts) during the image build
- To do this we will copy the JSON version file into the image using the COPY declarative in `Dockerfile` e.g.  `COPY ./es- releases.json /root`  ... see https://docs.docker.com/engine/reference/builder/#copy
- Next will invoke `setup-container.sh` just as before but during image build
- Now that we are doing it during image build we can add kafka back
- Finally we'll modify dockershell.sh such that `--es-version` is now used to pick the correct image TAG

File: container/Dockerfile [hunk 8]
```diff
@@ -1,10 +1,21 @@
...
+ENV IOT_REPO_PATH=/root/db2eventstore-IoT-Analytics
+ENV KAFKA_REPO_PATH=/root/db2eventstore-kafka 
+ENV SPARK_HOME="/spark_home"
+ENV PATH=$PATH:$SPARK_HOME/bin:$SPARK_HOME/sbin
+ARG ES_VERSION=latest
 ARG BRANCH=master
+ARG SETUP_AREA=${IOT_REPO_PATH}/container/setup
+
+
+#breakpoints
```
Comment (hroumani) [inline container/Dockerfile]: I'm sure these are just temporary for debugging but just a reminder to remove these prior to the delivery.

File: container/Dockerfile [hunk 9]
```diff
@@ -37,19 +48,19 @@ RUN echo -e 'startup_message off \nhardstatus on \nhardstatus alwayslastline \nv
 # Setup vim 
 RUN echo -e 'syntax enable \nset tabstop=4 \nset number \nset hlsearch \nset incsearch' > ${HOME}/.vimrc
 
-# Setup spark/scala/python/rest
-ENV SPARK_HOME="/spark_home"
-ENV PATH=$PATH:$SPARK_HOME/bin:$SPARK_HOME/sbin
+# Setup java/scala/python/kafka/rest
```
Comment (hroumani) [inline container/Dockerfile]: Will you delete all of the below ?  At this point it makes sense for all the "setup" to be driven by `setup-container.sh` so it's in a single place (whether versioning applies or not) ?

File: container/build.sh [hunk 10]
```diff
@@ -27,8 +26,8 @@ while [ -n "$1" ]; do
         usage >&2
         exit 0
         ;;
-    -t|--tag)
-        TAG="$2"
+    -es|--es-version)
```
Comment (hroumani) [inline container/build.sh:30]: I was debating if there would be any value in keeping `tag` as an additional option but I don't think so i.e. even if developers want to play around and make changes they can just use `--es-version` to produce an image with a different tag and that seems appropriate.

File: container/dockershell.sh [hunk 11]
```diff
@@ -57,6 +57,7 @@ OPTIONS:
                   e.g. "db2eventstore-1578174815082"
 --user         User name of the Event Store server
 --password     Password of the Event Store server
+--es-version   Event Store target cluster version
```
Comment (hroumani) [inline container/dockershell.sh]: I'd add a comment specifying that this is used to pick the tag of the image to search for, given the image is built with the version as the tag.

**Review by hroumani — CHANGES_REQUESTED** [review]
Looks great, I know it's still a work in progress but just a few comments.

Only high level suggestion is to move all the setup into `setup-container.sh` and out of `Dockerfile` so it's all contained in one file (whether it requires version specific setup or not).

File: container/Dockerfile (hunk 8 above)
Comment (robertpangborn) [reply container/Dockerfile]: done

File: container/Dockerfile [hunk 12]
```diff
@@ -1,10 +1,13 @@
...
-ENV IOT_REPO_PATH=/root/db2eventstore-IoT-Analytics \
-    KAFKA_REPO_PATH=/root/db2eventstore-kafka
-ARG SETUP_AREA=${IOT_REPO_PATH}/container/setup
+ENV IOT_REPO_PATH=/root/db2eventstore-IoT-Analytics
+ENV KAFKA_REPO_PATH=/root/db2eventstore-kafka 
+ENV SPARK_HOME="/spark_home"
+ENV PATH=$PATH:$SPARK_HOME/bin:$SPARK_HOME/sbin
+ARG ES_VERSION=latest
 ARG BRANCH=master
+ARG SETUP_AREA=${IOT_REPO_PATH}/container/setup
```
Comment (hroumani) [inline container/Dockerfile:10]: Why is `SETUP_AREA` and `ARG` ?  I don't believe it's passed in by anyone ?  i.e. should it not be an `ENV` like the rest above ?  

`BRANCH` and `ES_VERSION` make sense as those are args (with the default values above) passed in by `build.sh`.

File: container/Dockerfile [hunk 13]
```diff
@@ -37,19 +40,11 @@ RUN echo -e 'startup_message off \nhardstatus on \nhardstatus alwayslastline \nv
...
-RUN ${SETUP_AREA}/setup-kafka.sh
-RUN ${IOT_REPO_PATH}/rest/install.sh
-
-# Copy ES 2.0.1.0 client libraries
-RUN echo "Downloading Event Store 2.0.1.0 client libraries, it will take some time..."
-RUN wget -P ${SPARK_HOME}/jars/ https://repo1.maven.org/maven2/com/ibm/event/ibm-db2-eventstore-client-spark-2.4.6/2.0.1.0/ibm-db2-eventstore-client-spark-2.4.6-2.0.1.0.jar
+# copy json file to user_volume
+COPY ./es-releases.json /root/user_volume/
+
+# get variabls from json and setup sparke
```
Comment (hroumani) [inline container/Dockerfile:46]: This comment seems wrong ?  Can you just change it to:
`# Setup all container dependencies based on the selected version`

File: container/Dockerfile [hunk 14]
```diff
@@ -37,19 +40,11 @@ RUN echo -e 'startup_message off \nhardstatus on \nhardstatus alwayslastline \nv
...
-RUN ${SETUP_AREA}/setup-spark.sh
-RUN ${SETUP_AREA}/setup-scala.sh
-RUN ${SETUP_AREA}/setup-python.sh
-RUN ${SETUP_AREA}/setup-kafka.sh
-RUN ${IOT_REPO_PATH}/rest/install.sh
-
-# Copy ES 2.0.1.0 client libraries
-RUN echo "Downloading Event Store 2.0.1.0 client libraries, it will take some time..."
-RUN wget -P ${SPARK_HOME}/jars/ https://repo1.maven.org/maven2/com/ibm/event/ibm-db2-eventstore-client-spark-2.4.6/2.0.1.0/ibm-db2-eventstore-client-spark-2.4.6-2.0.1.0.jar
+# copy json file to user_volume
```
Comment (hroumani) [inline container/Dockerfile:43]: Can you amend this comment as this is an important line:

```
# Copy the json/version file to the user_volume directory so it can be read by setup-container.sh.
# It is important to note that the location of this file (i.e. es-release.json) is expected to be in the same
# directory as this script (i.e. build.sh) if that ever changes this will need to be updated.

File: container/README.md [hunk 15]
```diff
@@ -4,17 +4,18 @@ The Dockerfile in this directory will build a docker image named `eventstore_dem
 
 **Procedure:**
 **Step 1: Build the docker image**
-User need to run the shell script `build.sh` to build the docker image.
-The image size is around 3.3 GB, build takes around 30 mins, depending on the network condition.
+Run the shell script `build.sh` to build the docker image.
+The image size is around 3.3 GB, build takes around 30 mins, depending on network conditions.
+The Event Store release the IoT applications will use must be specified. The release is used to tag the image. Supported releases are: 2.0.1.0 and 2.0.0.5.
```
Comment (hroumani) [inline container/README.md:9]: NIT - "The Event Store release **that** the IOT..."

Comment (hroumani) [reply container/README.md:9]: Also this comment is **incorrect* no ?  It doesn't have to be specified i.e. it's not mandatory, we default to "latest", can you update this comment ?

File: container/dockershell.sh [hunk 16]
```diff
@@ -57,6 +57,7 @@ OPTIONS:
                   e.g. "db2eventstore-1578174815082"
 --user         User name of the Event Store server
 --password     Password of the Event Store server
+--es-version   Event Store target cluster version used to tag image
```
Comment (hroumani) [inline container/dockershell.sh:60]: Should add that if not specified we will assume the "latest" image tag which may not be correct if you built with an earlier release.

File: container/setup/setup-container.sh [hunk 17]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+# test if ES version supplied
+if [ -z $1 ]; then
```
Comment (hroumani) [inline container/setup/setup-container.sh:4]: Hmm, does this even work ?

I'm probably missing something but in `Dockerfile` if you don't specify a es_version we will set it to `LATEST` ->
```
ARG ES_VERSION=latest
```

So the string won't be empty in that case ?

File: container/setup/setup-container.sh [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+
+# set the version variables
+SPARK_VERSION=$(jq -r .release.${ES_VERSION}.sparkversion $USER_VOLUME/es-releases.json)
+SPARK_CLIENT=$(jq -r .release.${ES_VERSION}.sparkclient $USER_VOLUME/es-releases.json)
+
+# Setup java/scala/python/kafka/rest/spark
+${SETUP_AREA}/setup-java.sh
+${SETUP_AREA}/setup-scala.sh
+${SETUP_AREA}/setup-python.sh
+${SETUP_AREA}/setup-kafka.sh
```
Comment (hroumani) [inline container/setup/setup-container.sh:21]: I think more work is needed for kafka here, recall that the KAFKA repo actually stores a hard-coded es-client JAR i.e. here:
https://github.com/IBMProjectEventStore/db2eventstore-kafka/tree/master/event-stream/lib

We did add a tag for the older release:
https://github.com/IBMProjectEventStore/db2eventstore-kafka/tags

So `setup-kafka` would need to also be passed the correct version to clone.

FYI feel free to handle this as a *follow on* issue.

**Review by hroumani — CHANGES_REQUESTED** [review]
Nice work, it's going in the right direction.

Handful of questions/comments.

File: container/Dockerfile (hunk 12 above)
Comment (robertpangborn) [reply container/Dockerfile:10]: setup area was defined as ARG in the original script so I did not change it

File: container/setup/setup-container.sh (hunk 17 above)
Comment (robertpangborn) [reply container/setup/setup-container.sh:4]: In the docs I am saying the version has to be explicitly provided when running the build and that results it being passed down to setup container. I much prefer naming things explicitly and not relying on code to handle if values are not provided.

File: container/README.md (hunk 15 above)
Comment (robertpangborn) [reply container/README.md:9]: not sure grammatically "that" is required maybe a comma but will add that that

I know there is still some code testing for es version being passed and if not latest is used. I did not pull that code as there were many bits in play. I think by requiring the version it canshould be removed.

File: container/dockershell.sh (hunk 16 above)
Comment (robertpangborn) [reply container/dockershell.sh:60]: think we should be requiring version and relying on a defult version

File: container/setup/setup-container.sh (hunk 18 above)
Comment (robertpangborn) [reply container/setup/setup-container.sh:21]: I think with the sprint almost done it makes sense to make Kafka, and the other apps es version aware in a follow on story
